### PR TITLE
[Azure.Monitor.Ingestion] Add TroubleshootingGuide

### LINKED ISSUE #34612: [Monitor Ingestion] Add troubleshooting guide
As a prerequisite to CSS onboarding, create a troubleshooting guide that customers can reference as they encounter problems.

### PR DESCRIPTION
**Pull request by nisha-bhatia**
Fixes #34612
# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

File: sdk/monitor/Azure.Monitor.Ingestion/README.md [hunk 1]
```diff
@@ -253,7 +253,7 @@ Console.WriteLine("Table entry count: " +
 
 ### Enabling logging
```
Comment (scottaddie) [inline sdk/monitor/Azure.Monitor.Ingestion/README.md:255]: Remove this heading
```suggestion
```

File: sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
+# Troubleshooting Azure Monitor Ingestion client library issues
+
+This troubleshooting guide contains instructions to diagnose frequently encountered issues while using the Azure Monitor Ingestion client library for .Net.
```
Comment (scottaddie) [inline sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md:3]:
```suggestion
This troubleshooting guide contains instructions to diagnose frequently encountered issues while using the Azure Monitor Ingestion client library for .NET.
```

File: sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    * [Troubleshooting missing logs](#troubleshooting-missing-logs)
+    * [Troubleshooting slow logs upload](#troubleshooting-slow-logs-upload)
+
+## General troubleshooting
+
+### Enable client logging
+
+To troubleshoot issues with the library, first enable logging to monitor the behavior of the application. The errors and warnings in the logs generally provide useful insights into what went wrong and sometimes include corrective actions to fix issues.
+
+This library uses the standard [logging](https://docs.microsoft.com/dotnet/azure/sdk/logging) library. Basic information about HTTP sessions, such as URLs and headers, is logged at the `INFO` level.
```
Comment (scottaddie) [inline sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md:22]:
```suggestion
This library uses the standard [logging](https://learn.microsoft.com/dotnet/azure/sdk/logging) library. Basic information about HTTP sessions, such as URLs and headers, is logged at the `INFO` level.
```

File: sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+The simplest way to see the logs is to enable console logging. To create an Azure SDK log listener that outputs messages to the console, use the [AzureEventSourceListener.CreateConsoleLogger](https://docs.microsoft.com/dotnet/api/azure.core.diagnostics.azureeventsourcelistener.createconsolelogger?view=azure-dotnet) method:
```
Comment (scottaddie) [inline sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md:24]:
```suggestion
The simplest way to see the logs is to enable console logging. To create an Azure SDK log listener that outputs messages to the console, use the [AzureEventSourceListener.CreateConsoleLogger](https://learn.microsoft.com/dotnet/api/azure.core.diagnostics.azureeventsourcelistener.createconsolelogger?view=azure-dotnet) method:
```

File: sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+If you get an HTTP error with status code 403 (Forbidden), it means the provided credentials lack sufficient permissions to query the workspace.
+
+```text
+"{"error":{"message":"The provided credentials have insufficient access to perform the requested operation","code":"InsufficientAccessError","correlationId":""}}"
+```
+
+Confirm that:
+
+1. Sufficient permissions have been granted to the application or user making the request. For more information, see [manage access to workspaces](https://docs.microsoft.com/azure/azure-monitor/logs/manage-access#manage-access-using-workspace-permissions).
```
Comment (scottaddie) [inline sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md:45]:
```suggestion
1. Sufficient permissions have been granted to the application or user making the request. For more information, see [manage access to workspaces](https://learn.microsoft.com/azure/azure-monitor/logs/manage-access#manage-access-using-workspace-permissions).
```

File: sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+1. You're authenticating as that user or application if the user or application is granted sufficient privileges to query the workspace. If you're authenticating using the [DefaultAzureCredential](https://docs.microsoft.com/dotnet/api/azure.identity.defaultazurecredential?view=azure-dotnet), check the logs to verify the credential used is the one you expected. To enable logging, see the [Enable client logging](#enable-client-logging) section.
```
Comment (scottaddie) [inline sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md:46]:
```suggestion
1. You're authenticating as that user or application if the user or application is granted sufficient privileges to query the workspace. If you're authenticating using the [DefaultAzureCredential](https://learn.microsoft.com/dotnet/api/azure.identity.defaultazurecredential?view=azure-dotnet), check the logs to verify the credential used is the one you expected. To enable logging, see the [Enable client logging](#enable-client-logging) section.
```

File: sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+For more help with troubleshooting authentication errors, see the Azure Identity client library [troubleshooting guide](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/TROUBLESHOOTING.md).
+
+## Troubleshooting logs ingestion
+
+### Troubleshooting authorization errors
+
+If you get an HTTP error with status code 403 (Forbidden), it means that the provided credentials have insufficient permissions to upload logs to the specified Data Collection Endpoint (DCE) and Data Collection Rule (DCR) ID.
```
Comment (scottaddie) [inline sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md:54]:
```suggestion
If you get an error with HTTP status code 403 (Forbidden), it means that the provided credentials have insufficient permissions to upload logs to the specified Data Collection Endpoint (DCE) and Data Collection Rule (DCR) ID.
```

File: sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+com.azure.core.exception.HttpResponseException: Status code 403, "{"error":{"code":"OperationFailed","message":"The 
+authentication token provided does not have access to ingest data for the data collection rule with immutable Id 
+'<REDACTED>' PipelineAccessResult: AccessGranted: False, IsRbacPresent: False, IsDcrDceBindingValid: , DcrArmId: <REDACTED>,
+ Message: Required authorization action was not found for tenantId <REDACTED> objectId <REDACTED> on resourceId <REDACTED>
+ ConfigurationId: <REDACTED>.."}}"
+```
+
+1. Check that the application or user making the request has sufficient permissions:
+   * See this document to [manage access to data collection rule](https://learn.microsoft.com/azure/azure-monitor/logs/tutorial-logs-ingestion-portal#assign-permissions-to-the-dcr).
+   * To ingest logs, ensure the service principal is assigned the **Monitoring Metrics Publisher** role for the data collection rule.
```
Comment (scottaddie) [inline sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md:66]:
```suggestion
   * To ingest logs, ensure the service principal is assigned the **Monitoring Metrics Publisher** role for the DCR.
```

File: sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+1. If the user or application is granted sufficient privileges to upload logs, ensure you're authenticating as that user/application. If you're authenticating using the [DefaultAzureCredential](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/README.md#defaultazurecredential), check the logs to verify that the credential used is the one you expected. To enable logging, see the [Enable client logging](#enable-client-logging) section.
+1. The permissions may take up to 30 minutes to propagate. So, if the permissions were granted recently, retry after some time.
```
Comment (scottaddie) [inline sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md:68]:
```suggestion
1. The permissions may take up to 30 minutes to propagate. If the permissions were granted recently, retry after some time.
```

File: sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+### Troubleshooting missing logs
+
+When you send logs to Azure Monitor for ingestion, the request may succeed, but you may not see the data appearing in the designated Log Analytics workspace table as configured in the DCR. To investigate and resolve this issue, ensure the following:
+
+* Double-check that you're using the correct data collection endpoint when configuring the `LogsIngestionClient`. Using the wrong endpoint can result in data not being properly sent to the Log Analytics workspace.
```
Comment (scottaddie) [inline sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md:74]:
```suggestion
* Double-check that you're using the correct DCE when configuring the `LogsIngestionClient`. Using the wrong endpoint can result in data not being properly sent to the Log Analytics workspace.
```

File: sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+* Keep in mind that data ingestion may take some time, especially if you're sending data to a specific table for the first time. In such cases, allow up to 15 minutes for the data to be fully ingested and available for querying and analysis.
+
+### Troubleshooting slow logs upload
+
+If you experience delays when uploading logs, it could be due to reaching service limits, which may trigger the rate limiter to throttle your client. To determine if your client has reached service limits, you can enable logging and check if the service is returning errors with an HTTP status code 429. For more information on service limits, see the [Azure Monitor service limits documentation](https://learn.microsoft.com/azure/azure-monitor/service-limits#logs-ingestion-api).
+
+To enable client logging and to troubleshoot this issue further, see the instructions provided in the section titled [Enable client logging](#enable-client-logging).
+
+If there are no throttling errors, then consider increasing the concurrency to upload multiple log requests in parallel.
+To set the concurrency, use the `UploadLogsOptions` type's `MaxConcurrency` property.
```
Comment (scottaddie) [inline sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md:91]:
```suggestion
To set the concurrency, use the `UploadLogsOptions.MaxConcurrency` property.
```

File: sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+For more help with troubleshooting authentication errors, see the Azure Identity client library [troubleshooting guide](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/TROUBLESHOOTING.md).
+
+## Troubleshooting logs ingestion
+
+### Troubleshooting authorization errors
+
+If you get an HTTP error with status code 403 (Forbidden), it means that the provided credentials have insufficient permissions to upload logs to the specified Data Collection Endpoint (DCE) and Data Collection Rule (DCR) ID.
+
+```text
+com.azure.core.exception.HttpResponseException: Status code 403, "{"error":{"code":"OperationFailed","message":"The
```
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md:57]: This looks like a Java stack trace. Can we add the .NET one?

File: sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+### Troubleshooting slow logs upload
+
+If you experience delays when uploading logs, it could be due to reaching service limits, which may trigger the rate limiter to throttle your client. To determine if your client has reached service limits, you can enable logging and check if the service is returning errors with an HTTP status code 429. For more information on service limits, see the [Azure Monitor service limits documentation](https://learn.microsoft.com/azure/azure-monitor/service-limits#logs-ingestion-api).
+
+To enable client logging and to troubleshoot this issue further, see the instructions provided in the section titled [Enable client logging](#enable-client-logging).
+
+If there are no throttling errors, then consider increasing the concurrency to upload multiple log requests in parallel.
+To set the concurrency, use the `UploadLogsOptions` type's `MaxConcurrency` property.
+
+```C# Snippet:UploadWithMaxConcurrency
```
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md:93]: I think we are missing the `.cs` file where this snippet is defined.

File: sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+    Uri endpoint = new Uri("<data_collection_endpoint_uri>");
+    string ruleId = "<data_collection_rule_id>";
+    string streamName = "<stream_name>";
+
+    var credential = new DefaultAzureCredential();
+    LogsIngestionClient client = new(endpoint, credential);
```
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md:99]: nit: in all other cases in this snippet we use `new FooType()`. I think this is a bit more typical for samples. I would suggest this pattern:
```c#
var foo = new Bar();
```

File: sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+
+    DateTimeOffset currentTime = DateTimeOffset.UtcNow;
+
+    var entries = new List<Object>();
```
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md:103]:
```suggestion
    var entries = new List<object>();
```

File: sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+        entries.Add(
+            new {
+                Time = currentTime,
+                Computer = "Computer" + i.ToString(),
+                AdditionalContext = i
+            }
+        );
+    }
+    // Set concurrency in LogsUploadOptions
+    LogsUploadOptions options = new LogsUploadOptions();
```
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md:115]: nit: use object initializer pattern

File: sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md (hunk 7 above)
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md:54]: Is it worth mentioning the specific exception? I think it is RequestFailedException in this case.

File: sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+   * To ingest logs, ensure the service principal is assigned the **Monitoring Metrics Publisher** role for the data collection rule.
+1. If the user or application is granted sufficient privileges to upload logs, ensure you're authenticating as that user/application. If you're authenticating using the [DefaultAzureCredential](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/README.md#defaultazurecredential), check the logs to verify that the credential used is the one you expected. To enable logging, see the [Enable client logging](#enable-client-logging) section.
+1. The permissions may take up to 30 minutes to propagate. So, if the permissions were granted recently, retry after some time.
+
+### Troubleshooting missing logs
+
+When you send logs to Azure Monitor for ingestion, the request may succeed, but you may not see the data appearing in the designated Log Analytics workspace table as configured in the DCR. To investigate and resolve this issue, ensure the following:
```
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md:72]:
```suggestion
When you send logs to Azure Monitor for ingestion, the request may succeed, but you may not see the data appear in the designated Log Analytics workspace table as configured in the DCR. To investigate and resolve this issue, ensure the following:
```

File: sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md (hunk 12 above)
Comment (nisha-bhatia) [reply sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md:57]: Yes, I will update this

File: sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+```text
+com.azure.core.exception.HttpResponseException: Status code 403, "{"error":{"code":"OperationFailed","message":"The 
+authentication token provided does not have access to ingest data for the data collection rule with immutable Id 
+'<REDACTED>' PipelineAccessResult: AccessGranted: False, IsRbacPresent: False, IsDcrDceBindingValid: , DcrArmId: <REDACTED>,
+ Message: Required authorization action was not found for tenantId <REDACTED> objectId <REDACTED> on resourceId <REDACTED>
+ ConfigurationId: <REDACTED>.."}}"
+```
+
+1. Check that the application or user making the request has sufficient permissions:
+   * See this document to [manage access to data collection rule](https://learn.microsoft.com/azure/azure-monitor/logs/tutorial-logs-ingestion-portal#assign-permissions-to-the-dcr).
```
Comment (scottaddie) [inline sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md:65]:
```suggestion
   * See this document to [manage access to DCR](https://learn.microsoft.com/azure/azure-monitor/logs/tutorial-logs-ingestion-portal#assign-permissions-to-the-dcr).
```

File: sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+
+```csharp
+using Azure.Core.Diagnostics;
+
+// set up a listener to monitor logged events
+using AzureEventSourceListener listener = AzureEventSourceListener.CreateConsoleLogger();
+```
+
+### Authentication errors
+
```
Comment (scottaddie) [inline sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md:34]: Missing content for this section

File: sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+### Troubleshooting insufficient access error for metrics query
```
Comment (scottaddie) [inline sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md:35]: Why is there a section for the Monitor Query library?

File: sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md [hunk 21]
```diff
@@ -0,0 +1,120 @@
+# Troubleshooting Azure Monitor Ingestion client library issues
+
+This troubleshooting guide contains instructions to diagnose frequently encountered issues while using the Azure Monitor Ingestion client library for .NET.
+
+## Table of contents
+
+* [General troubleshooting](#general-troubleshooting)
+    * [Enable client logging](#enable-client-logging)
+    * [Troubleshooting authentication issues](#authentication-errors)
+    * [Troubleshooting NoSuchMethodError or NoClassDefFoundError](#dependency-conflicts)
```
Comment (scottaddie) [inline sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md:10]: I don't see this section below

File: sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+If there are no throttling errors, then consider increasing the concurrency to upload multiple log requests in parallel.
+To set the concurrency, use the `UploadLogsOptions.MaxConcurrency` property.
+
+```C# Snippet:UploadWithMaxConcurrency
+var endpoint = new Uri("<data_collection_endpoint_uri>");
+var ruleId = "<data_collection_rule_id>";
+var streamName = "<stream_name>";
+
+var credential = new DefaultAzureCredential();
+LogsIngestionClient client = new(endpoint, credential);
```
Comment (scottaddie) [inline sdk/monitor/Azure.Monitor.Ingestion/TROUBLESHOOTING.md:80]: Let's be consistent with the object creation syntax in these code samples. Note the 2 different approaches taken in lines 79 and 80.

**Review by JoshLove-msft — APPROVED** [review]

**Pull request merged**